### PR TITLE
Fix static extents assignment for LayoutLeft/LayoutRight assignment

### PR DESCRIPTION
**Pull request by crtrott**
We always supported the following:
```c++
Kokkos::View<int[5]> a("A");
Kokkos::View<int[5]> b("A");
Kokkos::View<int*> c("C",5);
a = b;
b = a;
b = c;
```

We also supported for always:
```c++
Kokkos::View<int*,Kokkos::LayoutLeft> a("A",5);
Kokkos::View<int*,Kokkos::LayoutRight> b("B",5);
a = b;
b = a;
```

What apparently doesn't work is:
```c++
Kokkos::View<int[5],Kokkos::LayoutLeft> a("A");
Kokkos::View<int[5],Kokkos::LayoutRight> b("B");
a = b;
b = a;
```

Turns out we do some extra checks for dynamic rank match, when accounting for LayoutLeft/LayoutRight convertibility of Rank 0 and Rank 1 Views.
That check is unnecessary, since it will get checked via the dimension assignability check already.
I.e. we hadn't properly orthogonalized the extents assignability from the Layout assignability for mixing LayoutLeft/LayoutRight.

This fixes this by getting rid of the erroneous condition.

### CODE REVIEW

**Comment by crtrott** [comment]
Note the proper dimension check happens around line 281 or so.

**Comment by crtrott** [comment]
Fixes #5534 @lucbv

File: core/src/impl/Kokkos_ViewMapping.hpp [hunk 1]
```diff
@@ -1128,9 +1128,8 @@ struct ViewOffset<
   KOKKOS_INLINE_FUNCTION constexpr ViewOffset(
       const ViewOffset<DimRHS, Kokkos::LayoutRight, void>& rhs)
       : m_dim(rhs.m_dim.N0, 0, 0, 0, 0, 0, 0, 0) {
-    static_assert((DimRHS::rank == 0 && dimension_type::rank == 0) ||
-                      (DimRHS::rank == 1 && dimension_type::rank == 1 &&
-                       dimension_type::rank_dynamic == 1),
```
Comment (crtrott) [inline core/src/impl/Kokkos_ViewMapping.hpp:1133]: That line 1133 is the offending check.

**Comment by crtrott** [comment]
Note this behavior was always intended (as the fact that it works for matching layouts attest too) and this is also how mdspan works.

**Review by lucbv — APPROVED** [review]
I talked this through with @crtrott and agree with the logic

File: core/unit_test/TestViewIsAssignable.hpp [hunk 2]
```diff
@@ -94,6 +94,14 @@ TEST(TEST_CATEGORY, view_is_assignable) {
   // Layout assignment
   Impl::TestAssignability<View<int*, left, d_exec>,
                           View<int*, right, d_exec>>::test(true, true, 10);
+  Impl::TestAssignability<View<int[5], left, d_exec>,
+                          View<int*, right, d_exec>>::test(false, false, 10);
+  Impl::TestAssignability<View<int[10], left, d_exec>,
+                          View<int*, right, d_exec>>::test(false, true, 10);
+  Impl::TestAssignability<View<int*, left, d_exec>,
+                          View<int[5], right, d_exec>>::test(true, true, 5);
```
Comment (dalg24) [inline core/unit_test/TestViewIsAssignable.hpp:102]: Wait why do you pass `5`?  Isn't it something we only relaxed not so long ago?  https://github.com/kokkos/kokkos-core-wiki/pull/108
Please drop it here and below.

**Review by dalg24 — COMMENTED** [review]
The view assignable unit test could use some refactor (not requesting changes just saying)
Your changes should enable rank-0 view assignments as well isn't it?  If I am not mistake please add tests for it.

File: core/unit_test/TestViewIsAssignable.hpp (hunk 2 above)
Comment (crtrott) [reply core/unit_test/TestViewIsAssignable.hpp:102]: yeah unnescessary here, gonna change that.

Comment (crtrott) [reply core/unit_test/TestViewIsAssignable.hpp]: Also added the rank-0 mixed layout case.

File: core/unit_test/TestViewIsAssignable.hpp [hunk 3]
```diff
@@ -92,8 +92,18 @@ TEST(TEST_CATEGORY, view_is_assignable) {
                           View<double*, left, d_exec>>::test(false, false, 10);
 
   // Layout assignment
+  Impl::TestAssignability<View<int, left, d_exec>,
+                          View<int, right, d_exec>>::test(true, true);
```
Comment (dalg24) [inline core/unit_test/TestViewIsAssignable.hpp:96]: No right -> left?

**Review by dalg24 — APPROVED** [review]

File: core/unit_test/TestViewIsAssignable.hpp (hunk 3 above)
Comment (crtrott) [reply core/unit_test/TestViewIsAssignable.hpp:96]: I guess we could add them everywhere, but mostly this is about mismatch, not any specific layouts.

**Pull request merged**